Bjorncs/capabilities [run-systemtest]

bjorncs opened the pull request:
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

bjorncs committed:
Add environment variable for capabilities enforcement mode

bjorncs committed:
Introduce per method request access filtering to JRT

bjorncs committed:
Remove empty package

bjorncs committed:
Merge Java package 'c.y.s.tls.{auth,json,policy}' into 'c.y.s.tls'

Facilitate improved encapsulation of Vespa mTLS related classes

bjorncs committed:
Add 'CapabilitySet.has()' methods

bjorncs committed:
Move generic crypto helpers from 'c.y.s.tls' to 'c.y.s'

bjorncs committed:
Add 'X509CertificateUtils.getSubjectCommonName()'

bjorncs committed:
Simplify type definition for subject alternative names

bjorncs committed:
Add to-string helper to ConnectionAuthContext

bjorncs committed:
Implement RequireCapabilitiesFilter in jrt + misc

Add peerSpec to Target/Connection. Always provide ConnectionAuthContext.
Add helper for creating default, all-granting ConnectionAuthContext.

vekterli commented on jrt/tests/com/yahoo/jrt/InvokeErrorTest.java:
```diff
@@ -157,4 +160,36 @@ public void testConnectionError() {
+        assertFalse(filter.invoked);
+        target.invokeSync(r, timeout);
+        assertTrue(r.isError());
+        assertTrue(filter.invoked);
```
Should we add an explicit check that the `test` method itself is _not_ invoked when the filter fails? (and vice versa in the case where the RPC passes the authz filter)

vekterli commented on security-utils/src/main/java/com/yahoo/security/tls/PeerAuthorizer.java:
```diff
@@ -35,9 +35,7 @@ public PeerAuthorizer(AuthorizedPeers authorizedPeers) {
-        if (authorizedPeers.isEmpty()) {
-            return new ConnectionAuthContext(certChain, CapabilitySet.all(), Set.of());
-        }
+        if (authorizedPeers.isEmpty()) return ConnectionAuthContext.defaultAllCapabilities();
```
Consider if we also want to expose the peer certificate information in `ConnectionAuthContext` even if it has no authz rules and is implicitly granted all capabilities

vekterli reviewed (approved):
:+1:

bjorncs committed:
Improve testing of JRT access filtering

Test that method is not invoked when request is rejected by filter.
Move testing of filter accepting request to async- and sync-specific test classes.

bjorncs committed:
Include client certificate chain even when authorization is disabled

vekterli reviewed (approved):
:+1:
